Add inline comments to reader post cards.

Addison-Stavlo opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to # pe7F0s-13p-p2

## Proposed Changes

* Shows the most recent comment by default below reader post cards, with the input to add comments. 
* A toggle to "view more comments" / "view less comments" to collapse/expand comments beyond the most recent.
* Another link to the full post page - "view more comments on full post" - if we have expanded and are showing less than all comments.
* Adds stats and tracks for new features (expanding to see more comments, and clicking to view all comments on the post). updates tracks for clicking reply, cancel reply, and submitting a comment to have an `is_inline_comment` property.

![inline-comments](https://github.com/Automattic/wp-calypso/assets/28742426/1742a60a-4a7b-4692-820a-a8b9741ba17e)
(note - we decided to get rid of the loading state in the above GIF)

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Run this branch and test inline comments in various reader streams.
* Verify tracking events for reply/cancel buttons and posting a comment (and nested replies) have an `is_inline_comment` property added. Verify these dont happen in the comments section of the individual posts page.
* Verify tracking event exists for clicking to expand the comments section, and for clicking to "view all comments on post" (only happens on posts with many comments - eg Discover).

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

Addison-Stavlo committed:
add postcomments component to reader post card if post comments are open

Addison-Stavlo committed:
rebase trunk

Addison-Stavlo committed:
optionally display only most recent comment in comment list component

Addison-Stavlo committed:
add toggle to expand beyond most recent comment for post comment list

Addison-Stavlo committed:
fix reply button on inline comments and apply better pagination

Addison-Stavlo committed:
fix crash edge case of finding mostRecentComment for comment list

Addison-Stavlo committed:
only show expand toggle button if there is more than 1 comment

Addison-Stavlo committed:
only fetch comments if they arent in state, and dont show them on compact cards

Addison-Stavlo committed:
prevent event bubble when clicking comment area

Addison-Stavlo committed:
add a loading state to comments area

Addison-Stavlo committed:
hide comment count and load more pagination

Addison-Stavlo committed:
dont show comment form on logged out view

Addison-Stavlo committed:
update initial size and depth, add comment count

Addison-Stavlo committed:
add link to view all comments on the posts page

Addison-Stavlo committed:
return early if we cant show a comment form or any comments, fix edge case with view post callback noop

Addison-Stavlo committed:
have view all comments on post open the post at the comments section

roo2 commented:
Currently it looks like the comment is disconnected from the post visually. From the mockups, we need to take out the line below a post when there is a comment present. 
<img width="649" alt="Screen Shot 2023-08-10 at 12 16 34 pm" src="https://github.com/Automattic/wp-calypso/assets/22446385/40637e7b-a95d-4a4f-8ecb-3c2cab833cc4">


![comments](https://github.com/Automattic/wp-calypso/assets/22446385/d480285b-84de-4608-859f-6c9aac23c6a8)

roo2 commented:
The "reply" buttons should be hidden for logged out users
 
<img width="635" alt="Screen Shot 2023-08-10 at 12 19 34 pm" src="https://github.com/Automattic/wp-calypso/assets/22446385/a1cd513b-3eb8-4422-a1f9-f86bee9ea2b0">

roo2 commented:
> Once expanded and there are still more comments than we are showing, do we want the "view all comments" link to the post page comments section - OR do we want to enable expanding loading more comments inline in the posts form? The functionality is there, but it may be weird to expand out to 100+ comments in the reader stream post cards. I assumed the former for now...

Yea I'd assume the same, to link them to the full post page, it would be ideal too if when you land there, all comments are expanded.

roo2 commented:
It looks like there's still being space reserved when comments are disabled e.g.
<img width="664" alt="Screen Shot 2023-08-10 at 2 52 15 pm" src="https://github.com/Automattic/wp-calypso/assets/22446385/7ccde8da-9e8c-4a83-979b-099b561c71cc">

roo2 commented:
the way that the first reply to a comment is handled before reloading the page is not quite right e.g.
Create the first comment on a post, reply to it.

I would expect that both comments remain visible, but it seems that first only the parent comment is visible and then only the descendant comment

https://github.com/Automattic/wp-calypso/assets/22446385/49112aed-7818-42fa-8606-42c63cb65115

roo2 commented on client/blocks/comments/post-comment-list.jsx:
```diff
@@ -328,6 +387,23 @@ class PostCommentList extends Component {
+			// means when we expand from only seeing the most recent found here, it may not be
+			// rendered in the longer list if it is a reply but not a reply to one of the most
+			// recent top level comments. Do we find the top level comment corresponding to this
+			// most recent and push it into displayedComments? Do we want to redo how
```
I'm not sure if this is exactly the same question that you raised in slack, but I think we should definitely display all of the ancestor comments if we can for context, otherwise the comment may not make sense or even have the opposite meaning.

Addison-Stavlo replied:
Yes this is noting the same thing from the conversation in slack. 😁 

>but I think we should definitely display all of the ancestor comments if we can for context, otherwise the comment may not make sense or even have the opposite meaning.

Talking about the expanded view, its not necessarily about showing all ancestor comments for context (its doing that already, rendering the full tree for those comments). But its doing so for the comment trees based off the recentness of the top-level comments and not the replies, which might not necessarily contain the most recent reply we show in the collapsed view (if it was replying on an older top-level comment). 

Combining this with what you noted above about the collapsed view feeling weird, there are a lot of options for what we can do here and what we actually want to show in each view. I imagine @davemart-in will have some input when he gets back as well so I probably wont dive too deep in changing these yet.

Addison-Stavlo commented:
Great catches, thanks! Il clean these up.

> Currently it looks like the comment is disconnected from the post visually. From the mockups, we need to take out the line below a post and reduce padding when there is a comment present.

I should have noted im ignoring a lot of the style changes here since the majority of those changes should be inherited when we include https://github.com/Automattic/wp-calypso/pull/80103. So im holding off on final css updates until we bring that in.

Addison-Stavlo committed:
update styles to make comments more cohesive with card

Addison-Stavlo committed:
hide reply and like buttons if logged out

Addison-Stavlo committed:
hide comments section if user cannot comment and there are no comments to show

Addison-Stavlo committed:
change view all lingo to view more

Addison-Stavlo commented:
Updated/fixed the following:
> * Currently it looks like the comment is disconnected from the post visually. 
>  * The "reply" buttons should be hidden for logged out users
> * It looks like there's still being space reserved when comments are disabled e.g.

Regarding:
> Yea I'd assume the same, to link them to the full post page, it would be ideal too if when you land there, all comments are expanded.

I changed the lingo to 'view more' instead of 'view all'. Considering we will have a limited list of comments inline, and popular posts can have a LOT of comments, opening to the post page to the default amount of comments it currently loads I think can make sense. It will be more comments than they saw inline, but not a giant jump to hundreds initially. And the extra comment loading button will be right there for them if needed:

<img width="576" alt="Screenshot 2023-08-10 at 9 40 50 AM" src="https://github.com/Automattic/wp-calypso/assets/28742426/a13a6ce8-8daa-4e40-a85d-9ca9f3fe427b">

Regarding
>  the way that the second comment on a post is handled is a bit strange. ... The behaviour is slightly different if the second comment is a reply. 

It is strange, but im not sure I entirely follow how its 'slightly different' for replies? We are showing the most recent comment/reply regardless so as soon as a new one is posted that is the only one that is shown.

That said, one idea could be - IF the most recent comment is a reply to another comment... show 2 comments? 🤔  (The reply and what it was replying to).

Addison-Stavlo committed:
refactor postCardComments to be its own component

Addison-Stavlo committed:
add tracks and stats

Addison-Stavlo committed:
Merge branch 'trunk' into add/comments-to-reader-post-cards

roo2 commented:
> IF the most recent comment is a reply to another comment... show both? (the reply and what it was replying to). The context of the most recent reply will make more sense, the context of what its replying to may or may not be there depending on its nested level in the conversation tree... but it seems like a step up from where we are now. 🤔

It is complicated hey.

I think there are two different situations:
* One is an uninvolved user who just scrolls to the post.
* The other is a user who has just made a comment.

I think if a user has just made a comment, we definitely want to show their new comment.
But if a new user is just scrolling down the page, we don't want to show them a single nested reply without context.
Maybe if the user has made a comment, we should show it along with all of its ancestors, but if a user has not made a comment we should just show the latest top level comment?

Does that sound like a good way to handle it to you?

Addison-Stavlo commented:
> Does that sound like a good way to handle it to you?

Possibly!  I think the important thing you are highlighting is that we could have different behavior between users who have and have not commented on the post, i did not think of that before but I think it makes a lot of sense!

davemart-in commented:
A couple of things I spotted:

1. I'm not sure if it was just because it was the first time I was loading comments for this PR, but this error popped up a number of times:

![CleanShot 2023-08-16 at 15 18 59](https://github.com/Automattic/wp-calypso/assets/5634774/9817e69c-51f8-40c0-b4eb-902e47ad706c)

Wasn't sure if you had seen similar behavior.

2. On /read/conversations and /read/conversations/a8c when you click into the comment box, the page is redirected. I expected to be able to just comment inline:

![CleanShot 2023-08-16 at 15 22 30](https://github.com/Automattic/wp-calypso/assets/5634774/bb1fd83e-9987-49e5-a681-609dde7a2584)

3. I noticed on the logged out screens we're not showing the comment box. Thoughts on just showing the comment box and treating it like any of the other logged in actions (redirect them to the login page) when someone clicks in?

davemart-in commented:
I wonder if we just fail silently for this instead of showing an error:

![image](https://github.com/Automattic/wp-calypso/assets/5634774/ce7818e3-a8a8-463d-8b8f-4ae6cc108a4a)

What do you think @Addison-Stavlo and @roo2?

davemart-in commented:
I also just saw that Eru fixed the second item above with https://github.com/Automattic/wp-calypso/pull/80603

roo2 commented:
> I wonder if we just fail silently for this instead of showing an error:

> What do you think @Addison-Stavlo and @roo2?

I think we'd better investigate and try to work out why it's failing first, maybe there's a certain type of error that we can ignore or it might be that there's some bug we need to fix

Addison-Stavlo commented:
> this error popped up a number of times:
 I;ve seen it a number of times now in testing, and I'm not sure how helpful it is to the user.

I actually have not been encountering this at all. I just built the branch and have been navigating around different streams and am not encountering it. Is there anywhere specific this is popping up for you @davemart-in ?

Addison-Stavlo commented:
> I noticed on the logged out screens we're not showing the comment box. Thoughts on just showing the comment box and treating it like any of the other logged in actions (redirect them to the login page) when someone clicks in?

I don't have any issues with that. I imagine in that case we should show the nested reply/like etc. buttons as well?

davemart-in commented:
>  I imagine in that case we should show the nested reply/like etc. buttons as well?

Yeah, let's just keep the experiences the same, except on the logged out, when they click into a comment reply field, or click the reply link, we redirect them to the login screen (hopefully returning them back to the same spot once they've logged in).

> Is there anywhere specific this is popping up for you @davemart-in?

I'm seeing it pretty consistently on the discover page. It's hard to know details, as I'm not alerted as to what post it is erroring out on. The error message is generic and there are no JS errors showing.

I'm happy to jump on a call and put some `console.log` code in with you via screen share if you'd like to see if we can get to the bottom of it together.

Addison-Stavlo committed:
limit comments error to development environment

Addison-Stavlo committed:
show only the most recent top level comment in collapsed mode, and the most recent reply by the reader

Addison-Stavlo committed:
rebase trunk

Addison-Stavlo committed:
rebase trunk

Addison-Stavlo committed:
remove test for removed error notice

Addison-Stavlo committed:
remove placeholder state

Addison-Stavlo committed:
reintroduce reply and comments form to logged out view, but have them direct to signup

Addison-Stavlo committed:
reduce padding under comment field

Addison-Stavlo committed:
update logic for expand collapse toggle to fix bug

Addison-Stavlo committed:
update is_inline_comment prop on isInlineComment to work on the nested replys

Addison-Stavlo committed:
cleanup debris from previous commits

Addison-Stavlo committed:
add exception to card click handler to ingore comments section

Addison-Stavlo commented:
> I'm seeing it pretty consistently on the discover page. 

Following up on the retrieving comments error noted above. This happens for jetpack connected sites who have the json-api module turned off. The module is on by default and should be encouraged to remain on for most users, but its expected that this sort of behavior would not work when disabled. For this PR, we have decided to remove the error notice (outside of the development environment) as we do not think it makes sense to surface this to users in any case.

<hr /> 

Beyond that, this should be ready for final review and testing.

davemart-in commented:
So, I just noticed this. 

![image](https://github.com/Automattic/wp-calypso/assets/5634774/d4c34e98-f786-49bb-8a9c-20e740e91c1f)

Looks like a pingback is showing up as the latest comment? Here's the post URL: https://philosophyvia.photos/2023/08/15/peaceful%e2%98%ae%ef%b8%8f-soul-lifes-ultimate-goal/

Here's the post itself:

![image](https://github.com/Automattic/wp-calypso/assets/5634774/313044e4-9083-4e43-8988-385ede9a8737)

Can we exclude pingbacks from showing up as the default comment and probably exclude pingbacks from showing up at all within inline comments?

Else, here is what we see with the comments expanded:

![image](https://github.com/Automattic/wp-calypso/assets/5634774/117dd68c-5e1e-46d9-b4c8-7f3b5f3bfdcd)

Not super useful to see pingbacks here IMO

Addison-Stavlo committed:
remove pingbacks and trackbacks from inline comments, fix logic with showing expand/collapse toggle

Addison-Stavlo commented:
Pingbacks and Trackbacks should be removed from inline comments now. I encountered a few regressions with expand/collapse toggle and 'view more on full post' links while enabling this but I think I fixed them all.

Note - I added something that might be odd, but in the case that we are showing all the non-pingback comments in the collapsed view I made the "view all comments on full post" link appear. When there is 1 comment being shown this may make sense, but when there are 0 comments being shown (eg only comments are pingbacks) it may be weird?

I was thinking of the edge case where we have 0 or 1 comment + pingbacks/trackbacks. We will say (via the comment icon in the card) that there are a handful of comments but only be showing 0 or 1 with no 'view more' toggle.

Heres what they look like currently:

<img width="511" alt="Screenshot 2023-08-21 at 1 07 47 PM" src="https://github.com/Automattic/wp-calypso/assets/28742426/b4f947d9-8365-48a0-89d8-e2ad4f019a1c">

<img width="492" alt="Screenshot 2023-08-21 at 1 09 07 PM" src="https://github.com/Automattic/wp-calypso/assets/28742426/d39d5cef-94cc-4b95-bec2-a138ec04e978">

Thoughts @davemart-in ? Should we remove that link entirely in the case where we aren't showing any inline comments? Does showing it when we are showing comments but there are other pingbacks we are not make sense?

davemart-in reviewed (approved):
Tested everything again just now. This looks good to me!

> Thoughts @davemart-in ? Should we remove that link entirely in the case where we aren't showing any inline comments? Does showing it when we are showing comments but there are other pingbacks we are not make sense?

Seems pretty edge case so I'm not sure if it's a big deal either way. That said, happy to keep the link there. At least there is consistency around the link always being there, which I like.

After @roo2 or someone else has had a chance to review the code, I'd say ship it!

roo2 commented on client/state/data-layer/wpcom/comments/index.js:
```diff
@@ -120,7 +121,10 @@ export const announceFailure =
 			: translate( 'Could not retrieve comments for post' );
 
-		dispatch( errorNotice( error, { duration: 5000 } ) );
+		const environment = config( 'env_id' );
```
I think a future improvement would be to set some state that displays an error panel here. Looking at the code I think if comments fail to load for any reason it would either stay in the loading state or say "no comments" which is incorrect. Another question is whether certain types of errors should be displayed to the user wheras the issue with the jetpack sites should be hidden.

(I think it's fine to merge)

Addison-Stavlo replied:
> . Looking at the code I think if comments fail to load for any reason it would either stay in the loading state or say "no comments" which is incorrect. 

That is how the error behaves on the full post page currently iirc, incorrectly saying "no comments" in that section. For inline comments we just dont render the inline comments section. Theres definitely some room for future improvement especially in the post-comment-list loading (or lack of loading) state.

roo2 commented on client/blocks/reader-post-card/post-card-comments.jsx:
```diff
@@ -0,0 +1,64 @@
+	};
+
+	if ( shouldHideComments || ! hasFetchedComments ) {
+		return null;
```
as a future improvement I think `! hasFetchedComments` could return some kind of loading state

Addison-Stavlo replied:
We had a loading state originally, but I thought it felt heavy on design (even just the blue loading spinner + a "loading comments..." dialogue). I was wondering if we should try to update that similar to other placeholders we have for cards and the editor that have the grey loading placeholders, etc.  @davemart-in advised to get rid of the placeholder in general. I think the no-placeholder feels better than our last one, but perhaps a better loading state could still be a future improvement.

roo2 commented on client/blocks/reader-post-card/post-card-comments.jsx:
```diff
@@ -0,0 +1,64 @@
+			showConversationFollowButton={ false }
+			showNestingReplyArrow={ true }
+			initialSize={ 5 }
+			pageSize={ 15 }
```
hmmm the `pageSize={ 15 }` doesn't seem to be having an effect in the UI, I'm not sure if it's neccessary?

Addison-Stavlo replied:
Good catch! Its not necessary here now that we removed the pagination in this view and replaced it with a link to the comments section of the full post page. Removed the unnecessary prop.

roo2 commented on client/blocks/comments/post-comment-list.jsx:
```diff
@@ -363,6 +423,61 @@ class PostCommentList extends Component {
+		};
+
+		// Go through the children of the last comment to find replies by the current user.
+		const authorReplies = commentsTree[ lastComment ]?.children.filter( ( replyId ) => {
```
I think it would be better to search all comments for an author reply, not just the latest thread. I also think it would be better to show the comment thread with the latest reply rather than the latest top level comment.

Having said that I think that's just a nit pick/ opinion and it works well as is.

Addison-Stavlo replied:
Yeah i think there are a lot of options on how what we could show for expanded / collapsed views. 

> I think it would be better to search all comments for an author reply, not just the latest thread. 

I think this could make sense for the expanded view, so various comment threads that the author has replied to are shown together. For the collapsed mode we don't care so much about multiple threads since we only want to show a little.

>  I also think it would be better to show the comment thread with the latest reply rather than the latest top level comment.

Also something i think could make a lot of sense for the expanded view. As collapsed seems like it would be messy given latest replys could be very nested and we would be showing more comments/replies than wanted in collapsed mode just for them to have context.

I would be excited to work on some future improvements here for what we show to the user.  That said one challenge here could be that for very large posts we only have the latest section of the comment tree returned by default and currently request more through pagination. So we are only working with a small recent section of the comment tree here currently and dont have the information to accurately go through all the comments for most recent replies, all author replies, etc. We would either need to request all comments (which could get heavyfor a lot of highly popular inline comment posts with 1000s of comments each), or create some new endpoints on the backend to return them differently. 🤔

roo2 commented:
A future improvement would be to have the comment indicator update immediately when making a new comment 
<img width="629" alt="Screen Shot 2023-08-22 at 12 27 53 pm" src="https://github.com/Automattic/wp-calypso/assets/22446385/aabb25fc-26d1-46e5-8ac4-986b88098f6d">

roo2 reviewed (approved):
LGTM! I just found very minor/nitpick stuff

Addison-Stavlo committed:
remove unnecessary page size prop

Addison-Stavlo committed:
memoize commentFetchingStatus selector

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8796615

Thank you @Addison-Stavlo for including a screenshot in the description! This is really helpful for our translators.
